File: VCF_Analyzer_py/VCF_analyzer.py
import os.path
import vcfpy
import pandas as pd
import matplotlib.pyplot as plt
import multiprocessing
from concurrent.futures import ProcessPoolExecutor, as_completed

def process_chunk(chunk):
    """Processes a chunk of VCF records into a DataFrame"""
    return pd.DataFrame(chunk)

class VCFProcessor:
    def __init__(self, input_vcfs : list, clinvar_vcf=None, gene_db=None):
        self.input_vcfs = input_vcfs if isinstance(input_vcfs, list) else [input_vcfs]
        self.merged_vcf = "merged_output.vcf"
        self.clinvar_vcf = clinvar_vcf
        self.gene_db = gene_db
        self.df = pd.DataFrame() # DataFrame to store data

    def merge_vcfs(self):
        """Merges multiple VCF into a single one"""
        if len(self.input_vcfs) == 1:
            print("Only one file found, skipping merging.")
            self.merged_vcf = self.input_vcfs[0]
            return

        print("Merging Files...")
        vcf_readers = [vcfpy.Reader(open(f, 'r')) for f in self.input_vcfs if os.path.exists(f)]
        if not vcf_readers:
            raise FileNotFoundError("No valid VCF files found.")
        vcf_writer = vcfpy.Writer(open(self.merged_vcf, 'w'), vcf_readers[0])

        for reader in vcf_readers:
            for record in reader:
                vcf_writer.write_record(record)

        print(f"Merged VCF saved to {self.merged_vcf}")

    def vcf_to_dataframe_parallel(self, num_workers=4):
        """Converts VCF into a Pandas DataFrame."""
        print("Converting VCF to DataFrame...")
        chunk_size = 50000
        vcf_reader = vcfpy.Reader(open(self.merged_vcf, 'r'))
        records = []

        with ProcessPoolExecutor(max_workers=num_workers) as executor:
            futures = []
            chunk = []

            for record in vcf_reader:
                chunk.append({
                    "CHROM": record.CHROM,
                    "POS": record.POS,
                    "ID": record.ID,
                    "REF":record.REF,
                    "ALT": ",".join(map(str, record.ALT)),
                    "QUAL": record.QUAL,
                    "FILTER": record.FILTER,
                    "DP": record.INFO.get("DP", 0),
                    "INFO": str(record.INFO)
                })

                if len(chunk) >= chunk_size:
                    futures.append(executor.submit(process_chunk, chunk))
                    chunk = [] # reset chunk

            if chunk:
                futures.append(executor.submit(process_chunk, chunk))

            df_chunks = [future.result() for future in as_completed(futures)]

        # Concatenate chunks
        self.df = pd.concat(df_chunks, ignore_index=True)
        print(f"Converted {len(self.df)} variants into DataFrame.")

    def filter_variants(self, min_qual=30, min_dp=10):
        """Filters variants based on QUAL and DP fields."""
        print("Filtering variants...")
        before_count = len(self.df)
        self.df = self.df[(self.df["QUAL"] >= min_qual) & (self.df["DP"] >= min_dp)]
        print(f"Remaining variants after filtering: {len(self.df)}")

    def count_variant_types(self):
        """Counts SNPs and Indels"""
        print("Counting SNPs and Indels...")
        self.df["Variant_Type"] = self.df.apply(lambda row: "SNP" if len(row["REF"]) == 1 and all(len(alt) == 1 for alt in row["ALT"].split(",")) else "INDEL", axis=1)
        counts = self.df["Variant_Type"].value_counts()
        print(counts)
        return counts

    def plot_qual_distribution(self):
        """Plots the distribution of variant quality scores."""
        print("Plotting variant quality distribution...")
        plt.hist(self.df["QUAL"], bins=50, color='blue', alpha=0.7)
        plt.xlabel("Variant Quality Score (QUAL)")
        plt.ylabel("Frequency")
        plt.title("Distribution of Variant Quality Scores")
        plt.show()

    def find_clinvar_variants(self):
        """Identifies variants present in ClinVar."""
        if not self.clinvar_vcf:
            print("No ClinVar file provided.")
            return

        print("Searching for ClinVar variants...")
        clinvar_df = pd.read_csv(self.clinvar_vcf, sep="\t")
        clinvar_positions = set(zip(clinvar_df["CHROM"], clinvar_df["POS"]))

        self.df["ClinVar_Match"] = self.df.apply(lambda row: (row["CHROM"], row["POS"]) in clinvar_positions, axis=1)
        matches = self.df[self.df["ClinVar_Match"]]
        print(f"Found {len(matches)} ClinVar matches.")
        return matches

    def annotate_variants(self):
        """Annotates variants file gene names."""
        if not self.gene_db:
            print("No gene database provided.")
            return

        print("Annotating variants...")
        gene_df = pd.read_csv(self.gene_db, sep="\t")
        gene_dict = dict(zip(zip(gene_df["CHROM"], gene_df["POS"]), gene_df["GENE"]))

        self.df["GENE"] = self.df.apply(lambda row:gene_dict.get((row["CHROM"], row["POS"]), "Unknown"), axis=1)
        print("Annotation completed.")

    def save_dataframe(self):
        """Save DataFrame to CSV"""
        self.df.to_csv("processed_variants.csv", index=False)
        print("Saved processed variants to CSV.")

    def run_pipeline(self):
        """Runs full VCF processing pipeline"""
        print("Starting VCF processing pipeline...")

        self.merge_vcfs()
        self.vcf_to_dataframe_parallel(num_workers=multiprocessing.cpu_count())
        self.filter_variants(min_qual=30, min_dp=10)
        self.count_variant_types()
        self.plot_qual_distribution()
        self.find_clinvar_variants()
        self.annotate_variants()
        self.save_dataframe()

        print("VCF processing completed.")

# Run Pipeline
if __name__ == "__main__":
    input_vcf = ["/Users/giuse/Downloads/case584_variants.vcf"]
    clinvar_vcf = "clinvar.vcf"
    gene_db = "gene_database.txt"

    processor = VCFProcessor(input_vcf, clinvar_vcf, gene_db)
    processor.run_pipeline()

File: VCF_Analyzer_Java/src/main/java/VCF_Analyzer.java
import htsjdk.variant.vcf.VCFFileReader;
import htsjdk.variant.variantcontext.VariantContext;
import htsjdk.variant.vcf.VCFHeader;
import org.apache.commons.csv.*;



import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

public class VCF_Analyzer {
    private List<String> inputVcfs;
    private String mergedVcf = "merged_output.vcf";
    private String clinvarFile;
    private String geneDbFile;
    private List<Map<String, String>> variantData = new ArrayList<>();

    public VCF_Analyzer(List<String> inputVcfs, String clinvarFile, String geneDbFile){
        this.inputVcfs = inputVcfs;
        this.clinvarFile = clinvarFile;
        this.geneDbFile = geneDbFile;
    }

    // Merge Multiple VCFs into one
    public void mergeVcfs() throws IOException{
        if (inputVcfs.size() == 1){
            System.out.println("Only one found, skipping merging.");
            mergedVcf = inputVcfs.get(0);
            return;
        }

        System.out.println("Merging VCF files...");
        try (BufferedWriter writer = Files.newBufferedWriter(Paths.get(mergedVcf))){
            for (String vcf : inputVcfs){
                try (BufferedReader reader = Files.newBufferedReader(Paths.get(vcf))){
                    String line;
                    while ((line = reader.readLine()) != null){
                        if (!line.startsWith(("#")) || vcf.equals(inputVcfs.get(0))){
                            writer.write(line);
                            writer.newLine();
                        }
                    }
                }
            }
        }
        System.out.println("Merged VCF saved to " + mergedVcf);


    }

    // Convert VCF to Java List
    public void vcfToListParallel(int numThreads){
        System.out.println("Converting VCF to List in parallel...");
        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        List<Future<List<Map<String, String>>>> futures = new ArrayList<>();

        try (VCFFileReader vcfReader = new VCFFileReader(new File(mergedVcf), false)){
            VCFHeader header = vcfReader.getFileHeader();
            List<VariantContext> variants = new ArrayList<>();
            vcfReader.iterator().forEachRemaining(variants::add);

            int chunkSize = 50000;
            for (int i = 0; i < variants.size(); i += chunkSize){
                List<VariantContext> chunk = variants.subList(i, Math.min(i + chunkSize, variants.size()));
                futures.add(executor.submit(() -> processVariantChunk(chunk)));
            }

            for (Future<List<Map<String, String>>> future : futures){
                variantData.addAll(future.get());
            }
        } catch (Exception e) {
            e.printStackTrace();
        }finally {
            executor.shutdown();
        }

        System.out.println("Converted " + variantData.size() + " variants into List.");

    }

    private List<Map<String, String>> processVariantChunk(List<VariantContext> chunk){
        List<Map<String, String>> processedData = new ArrayList<>();
        for (VariantContext variant : chunk){
            Map<String, String> row = new HashMap<>();
            row.put("CHROM", variant.getContig());
            row.put("POS", String.valueOf(variant.getStart()));
            row.put("ID", variant.getID());
            row.put("REF", variant.getReference().getBaseString());
            row.put("ALT", variant.getAlternateAlleles().toString());
            row.put("QUAL", String.valueOf(variant.getPhredScaledQual()));
            row.put("FILTER", variant.getFilters().toString());
            processedData.add(row);
        }
        return processedData;
    }

    // Filter Variants
    public void filterVariants(double minQual){
        System.out.println("Filtering variants...");
        int beforeCount = variantData.size();
        variantData = variantData.stream()
                .filter(v -> Double.parseDouble(v.get("QUAL")) >= minQual)
                .collect(Collectors.toList());
        System.out.println("Remaining variants after filtering: " + variantData.size() + " (removed " + (beforeCount - variantData.size()) + ")");
    }

    // Count SNPs and Indels
    public void countVariantTypes(){
        System.out.println("Counting SNPs and Indels...");
        Map<String, Long> counts = variantData.stream()
                .collect(Collectors.groupingBy(v -> v.get("REF").length() == 1 && v.get("ALT").length() == 1 ? "SNP" : "INDEL", Collectors.counting()));

        counts.forEach((key, value) -> System.out.println(key + ": " + value));
    }

    // Find ClinVar Matches
    public void findClinVarVariants() throws IOException{
        System.out.println("Searching for ClinVar variants...");
        Set<String> clinvarPositions = new HashSet<>();

        try (Reader reader = Files.newBufferedReader(Paths.get(clinvarFile));
            CSVParser csvParser = new CSVParser(reader, CSVFormat.DEFAULT.withDelimiter('\t').withFirstRecordAsHeader())){
            for (CSVRecord record : csvParser){
                clinvarPositions.add(record.get("CHROM") + ":" + record.get("POS"));

            }
        }

        long matches = variantData.stream()
                .filter(v -> clinvarPositions.contains(v.get("CHROM") + ":" + v.get("POS")))
                .count();
        System.out.println("Found " + matches + " ClinVar matches.");
    }

    // Annotate Variants
    public void annotateVariants() throws IOException{
        System.out.println("Annotating variants...");
        Map<String, String> geneDict = new HashMap<>();

        try (Reader reader = Files.newBufferedReader(Paths.get(geneDbFile));
             CSVParser csvParser = new CSVParser(reader, CSVFormat.TDF.withFirstRecordAsHeader())) {
            for (CSVRecord record : csvParser){
                geneDict.put(record.get("CHROM") + ":" + record.get("POS"), record.get("GENE"));
            }
        }

        variantData.forEach(v -> v.put("GENE", geneDict.getOrDefault(v.get("CHROM") + ":" + v.get("POS"), "Unknown")));
        System.out.println("Annotation completed.");
    }

    // Save to CSV
    public void saveDataFrame() throws IOException{
        System.out.println("Saving to CSV...");
        try (BufferedWriter writer = Files.newBufferedWriter(Paths.get("processed_variants.csv"));
             CSVPrinter csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT.withHeader("CHROM", "POS", "ID", "REF","ALT", "QUAL", "FILTER", "GENE"))){
            for (Map<String, String> row : variantData) {
                csvPrinter.printRecord(row.values());
            }
        }
        System.out.println("Saved processed variants to CSV.");
    }

    public static void main(String[] args) throws IOException{
        VCF_Analyzer analyzer = new VCF_Analyzer(Arrays.asList("sample1.vcf", "sample2.vcf"), "clinvar.tsv", "gene_database.tsv");
        analyzer.mergeVcfs();
        analyzer.vcfToListParallel(Runtime.getRuntime().availableProcessors());
        analyzer.filterVariants(30.0);
        analyzer.countVariantTypes();
        analyzer.findClinVarVariants();
        analyzer.annotateVariants();
        analyzer.saveDataFrame();
        System.out.println("VCF processing completed.");
    }
}
